Release notes for 1.7 (#803)
cert-manager/website: content/en/docs/release-notes/release-notes-1.7.md

## Breaking Changes (You **MUST** read this before you upgrade!)

⚠ Following their deprecation in version 1.5, the cert-manager API versions v1alpha2, v1alpha3, and v1beta1 have been removed.

> You must ensure that all cert-manager custom resources are stored in etcd at version v1
> and that all cert-manager `CustomResourceDefinition`s have only v1 as the stored version.

Reviewer: ```suggestion
You must ensure that all cert-manager custom resources are stored in etcd at version v1
and that all cert-manager `CustomResourceDefinition`s have only v1 as the stored version
**before** upgrading.
```

Since release 1.7, `cmctl` can automatically migrate any deprecated API resources.

> Please download `cmctl-v1.7.0-beta.0` and read [Removing Deprecated API Resources]

Reviewer 2: suggestion: if these are the actual draft release notes for 1.7, should we refer to `cmctl-v1.7.0` rather than `cmctl-v1.7.0-beta.0`?

also, could we link to the release page (https://github.com/jetstack/cert-manager/releases/tag/v1.7.0-beta.0 / https://github.com/jetstack/cert-manager/releases/tag/v1.7.0)?

```suggestion
Please download `cmctl-v1.7.0` from [here](https://github.com/jetstack/cert-manager/releases/tag/v1.7.0) and read [Removing Deprecated API Resources]
```

for full instructions.

[Removing Deprecated API Resources]: https://cert-manager.io/docs/installation/upgrading/remove-deprecated-apis/

## Major Themes

### Additional Certificate Output Formats

`additionalOutputFormats` is a field on the Certificate `spec` that allows
specifying additional supplementary formats of issued certificates and their

> private key. There are currently two supported additional output formats:
> `CombinedPEM` and `DER`. Both output formats can be specified on the same
> Certificate.

Reviewer 2: suggestion:

```suggestion
private key. There are currently two supported additional output formats:
`CombinedPEM` (the PEM-encoded private key followed by the certificate chain)
and `DER` (the DER-encoded private key only). Any combination of output formats
can be requested for the same certificate.
```

Author: I should make this same change in `usage/certificate/#additional-certificate-output-formats`

Read [Additional Certificate Output Formats] for more details and

> thanks to [@seuf](https://github.com/seuf) for getting this across the line!

Reviewer 2: love the shoutout :ok_hand:

[Additional Certificate Output Formats]: ../../usage/certificate/#additional-certificate-output-formats

### Removal of Deprecated APIs

> In 1.7 the cert-manager API versions v1alpha2, v1alpha3, and v1beta1 have been removed from the custom resource definitions (CRDs).
> You will notice that the YAML manifest files are much smaller as a result.
> These APIs have been deprecated since 1.5.

Reviewer: ```suggestion
In 1.7 the cert-manager API versions v1alpha2, v1alpha3, and v1beta1, that were deprecated in 1.5,
have been removed from the custom resource definitions (CRDs).
As a result, you will notice that the YAML manifest files are much smaller.
```

In this release we have added a new sub-command to the cert-manager CLI (`cmctl upgrade migrate-api-version`),
which you SHOULD run BEFORE upgrading cert-manager to 1.7.

> Please read [Removing Deprecated API Resources] for full instructions.

Reviewer 2: suggestion: I think I'd prefer to have Removal of Deprecated APIs before anything else under "Major Themes", it feels like the most important bit to me for people who might be reading these notes. It does stutter a bit following on from the breaking changes section, though. Totally not a blocker, just a personal preference thing.

### Server-Side Apply

This is the first version of cert-manager which relies on [Server-Side Apply].

> We are using it to properly manage the annotations and labels on the TLS Secret.

Reviewer: ```suggestion
We use it to properly manage the annotations and labels on TLS secrets.
```

> For this reason cert-manager 1.7 requires at least Kubernetes 1.18.

Reviewer 2: suggestion: total non-blocker, just throwing it out there.

```suggestion
For this reason cert-manager 1.7 requires at least Kubernetes 1.18 (see
[Supported Releases](https://cert-manager.io/docs/installation/supported-releases/) for further compatibility details).
```

[Server-Side Apply]: https://kubernetes.io/docs/reference/using-api/server-side-apply/

### Configuration Files

In this release we introduce a new configuration file for the cert-manager-webhook.
Instead of configuring the webhook using command line flags,
you can now modify the webhook Deployment to mount a ConfigMap
containing a configuration file.
Read the [WebhookConfiguration Schema] for more information.

> In future releases we will introduce configuration files for the other cert-manager components: controller-manager and cainjector.

Reviewer 2: suggestion: don't we usually call it just "controller"?

```suggestion
In future releases we will introduce configuration files for the other cert-manager components:
the controller and the cainjector.
```

[WebhookConfiguration Schema]: https://cert-manager.io/next-docs/reference/api-docs/#webhook.config.cert-manager.io/v1alpha1.WebhookConfiguration

### Non-bazel Development

We aim to stop using `bazel` for building and testing cert-manager.
It is too difficult for new contributors and frustrates many of our regular contributors.
So we have been fixing the cert-manager Makefile and trying to make it possible to run the unit-tests
with `go test ./cmd/... ./internal/... ./pkg/...` etc.

## Community

Thanks again to all open-source contributors with commits in this release, including:

- [@Adphi](https://github.com/Adphi)
- [@devholic](https://github.com/devholic)
- [@johnwchadwick](https://github.com/johnwchadwick)
- [@jsoref](https://github.com/jsoref)
- [@jwenz723](https://github.com/jwenz723)
- [@seuf](https://github.com/seuf)
- [@thirdeyenick](https://github.com/thirdeyenick)

And thanks as usual to [coderanger](https://github.com/coderanger) for helping people

> out on the Slack `#cert-manager` channel; it's a huge help and much appreciated.

Reviewer 2: love it :ok_hand:

Reviewer: Should `#cert-manager` be a link to somewhere that helps people get to slack?

Author: Done.
